markets/storageadapter/dealpublisher.go

```diff
@@ -0,0 +1,294 @@
+	p.cancelWaitForMoreDeals = cancel
+
+	go func() {
+		timer := time.NewTimer(p.publishPeriod)
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+		case <-timer.C:
+			p.lk.Lock()
+			defer p.lk.Unlock()
+
+			// The timeout has expired so publish all pending deals
+			log.Infof("publish deals queue period of %s has expired, publishing deals", p.publishPeriod)
+			p.publishAllDeals()
+		}
+	}()
+}
+
+func (p *DealPublisher) publishAllDeals() {
+	// If the timeout hasn't yet been cancelled, cancel it
+	if p.cancelWaitForMoreDeals != nil {
+		p.cancelWaitForMoreDeals()
+		p.cancelWaitForMoreDeals = nil
+		p.publishPeriodStart = time.Time{}
+	}
+
+	// Filter out any deals that have been cancelled
+	p.filterCancelledDeals()
+	deals := p.pending[:]
+	p.pending = nil
+
+	// Send the publish message
+	go p.publishReady(deals)
+}
+
+func (p *DealPublisher) publishReady(ready []*pendingDeal) {
+	if len(ready) == 0 {
+		return
+	}
+
+	deals := make([]market2.ClientDealProposal, 0, len(ready))
+	for _, pd := range ready {
+		deals = append(deals, pd.deal)
+	}
+
+	// Send the publish message
+	msgCid, err := p.publishDealProposals(deals)
+
+	// Signal that each deal has been published
+	for _, pd := range ready {
+		pd := pd
+		go func() {
+			res := publishResult{
+				msgCid: msgCid,
+				err:    err,
+			}
+			select {
+			case <-p.ctx.Done():
+			case pd.Result <- res:
+			}
+		}()
+	}
+}
+
+// Sends the publish message
+func (p *DealPublisher) publishDealProposals(deals []market2.ClientDealProposal) (cid.Cid, error) {
+	log.Infof("publishing %d deals in publish deals queue with piece CIDs: %s", len(deals), pieceCids(deals))
+
+	provider := deals[0].Proposal.Provider
+	mi, err := p.api.StateMinerInfo(p.ctx, provider, types.EmptyTSK)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	params, err := actors.SerializeParams(&market2.PublishStorageDealsParams{
+		Deals: deals,
+	})
+
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("serializing PublishStorageDeals params failed: %w", err)
+	}
+
+	smsg, err := p.api.MpoolPushMessage(p.ctx, &types.Message{
+		To:     market.Address,
+		From:   mi.Worker,
+		Value:  types.NewInt(0),
+		Method: market.Methods.PublishStorageDeals,
+		Params: params,
+	}, p.publishSpec)
+
+	if err != nil {
+		return cid.Undef, err
+	}
+	return smsg.Cid(), nil
+}
+
+func pieceCids(deals []market2.ClientDealProposal) string {
+	cids := make([]string, 0, len(deals))
+	for _, dl := range deals {
+		cids = append(cids, dl.Proposal.PieceCID.String())
+	}
+	return strings.Join(cids, ", ")
+}
+
+// filter out deals that have been cancelled
+func (p *DealPublisher) filterCancelledDeals() {
+	i := 0
+	for _, pd := range p.pending {
+		if pd.ctx.Err() == nil {
+			p.pending[i] = pd
+			i++
+		}
+	}
+	p.pending = p.pending[:i]
+}
+
+type publishResult struct {
+	msgCid cid.Cid
+	err    error
+}
+
+type pendingDeal struct {
+	ctx    context.Context
+	deal   market2.ClientDealProposal
+	Result chan publishResult
+}
```

Comment:
Would move those definitions next to DealPublisher so you don't have to look for them